libcperciva/network/network_connect.c

```diff
@@ -179,16 +185,34 @@ network_connect(struct sock_addr * const * sas,
 {
 
 	/* Let network_connect_timeo handle this. */
-	return (network_connect_timeo(sas, NULL, callback, cookie));
+	return (network_connect_timeo(sas, NULL, NULL, callback, cookie));
 }
 
 /**
- * network_connect_timeo(sas, timeo, callback, cookie):
- * Behave as network_connect(), but wait a duration of at most ${timeo} for
+ * network_connect_bind(sas, sa_b, callback, cookie):
+ * Iterate through the addresses in ${sas}, attempting to create and connect
+ * a non-blocking socket after binding to ${sa_b}.  Once connected, invoke
+ * ${callback}(${cookie}, s) where s is the connected socket; upon fatal error
+ * or if there are no addresses remaining to attempt, invoke
+ * ${callback}(${cookie}, -1).  Return a cookie which can be passed to
+ * network_connect_cancel() in order to cancel the connection attempt.
+ */
+void *
+network_connect_bind(struct sock_addr * const * sas, struct sock_addr const * sa_b,
+    int (* callback)(void *, int), void * cookie)
+{
+
+	/* Let network_connect_timeo handle this. */
+	return (network_connect_timeo(sas, sa_b, NULL, callback, cookie));
+}
+
+/**
+ * network_connect_timeo(sas, sa_b, timeo, callback, cookie):
+ * Behave as network_connect_bind(), but wait a duration of at most ${timeo} for
  * each address which is being attempted.
  */
 void *
-network_connect_timeo(struct sock_addr * const * sas,
+network_connect_timeo(struct sock_addr * const * sas, struct sock_addr const * sa_b,
```

Comment:
Oh, I see now why you were adding functionality to `network_connect_timeo`.  Let's do this differently -- shunt the functionality to an internal `static void * network_connect_internal`, then have `network_connect`, `network_connect_bind`, `network_connect_timeo` all use that.

Reply:
Done, could you please check if I understood you correctly?